test_conformance/basic/test_work_item_functions.cpp

```diff
@@ -35,146 +39,552 @@ struct work_item_data
+    outData[ind].globalSize[0] = (uint)get_global_size(dimindx);
+    outData[ind].globalID[0] = (uint)get_global_id(dimindx);
+    outData[ind].localSize[0] = (uint)get_local_size(dimindx);
+    outData[ind].localID[0] = (uint)get_local_id(dimindx);
+    outData[ind].numGroups[0] = (uint)get_num_groups(dimindx);
+    outData[ind].groupID[0] = (uint)get_group_id(dimindx);
+    outData[ind].globalOffset[0] = (uint)get_global_offset(dimindx);
+)";
+
+const char *outOfRangeWorkItemKernelCodeExt20 =
+    R"(
+    uint dimindx=get_work_dim()+1;
+    outData[ind].globalSize[0] = (uint)get_global_size(dimindx);
+    outData[ind].globalID[0] = (uint)get_global_id(dimindx);
+    outData[ind].localSize[0] = (uint)get_local_size(dimindx);
+    outData[ind].localID[0] = (uint)get_local_id(dimindx);
+    outData[ind].numGroups[0] = (uint)get_num_groups(dimindx);
+    outData[ind].groupID[0] = (uint)get_group_id(dimindx);
+    outData[ind].globalOffset[0] = (uint)get_global_offset(dimindx);
+    outData[ind].enqueuedLocalSize[0] = (uint)get_enqueued_local_size(dimindx);
+)";
+
+
+const char *outOfRangeWorkItemHardcodedKernelCodeExt =
+    R"(
+    outData[ind].globalSize[0] = (uint)get_global_size(4);
+    outData[ind].globalID[0] = (uint)get_global_id(4);
+    outData[ind].localSize[0] = (uint)get_local_size(4);
+    outData[ind].localID[0] = (uint)get_local_id(4);
+    outData[ind].numGroups[0] = (uint)get_num_groups(4);
+    outData[ind].groupID[0] = (uint)get_group_id(4);
+)";
+
+const char *outOfRangeWorkItemHardcodedKernelCodeExt11 =
+    R"(
+    outData[ind].globalSize[0] = (uint)get_global_size(4);
+    outData[ind].globalID[0] = (uint)get_global_id(4);
+    outData[ind].localSize[0] = (uint)get_local_size(4);
+    outData[ind].localID[0] = (uint)get_local_id(4);
+    outData[ind].numGroups[0] = (uint)get_num_groups(4);
+    outData[ind].groupID[0] = (uint)get_group_id(4);
+    outData[ind].globalOffset[0] = (uint)get_global_offset(4);
+)";
+
+const char *outOfRangeWorkItemHardcodedKernelCodeExt20 =
+    R"(
+    outData[ind].globalSize[0] = (uint)get_global_size(4);
+    outData[ind].globalID[0] = (uint)get_global_id(4);
+    outData[ind].localSize[0] = (uint)get_local_size(4);
+    outData[ind].localID[0] = (uint)get_local_id(4);
+    outData[ind].numGroups[0] = (uint)get_num_groups(4);
+    outData[ind].groupID[0] = (uint)get_group_id(4);
+    outData[ind].globalOffset[0] = (uint)get_global_offset(4);
+    outData[ind].enqueuedLocalSize[0] = (uint)get_enqueued_local_size(4);
+)";
+
+
+struct TestWorkItemBase
 {
-    int error;
+    TestWorkItemBase(cl_device_id did, cl_context c, cl_command_queue q,
+                     cl_uint data_size, const char *src)
+        : device(did), context(c), queue(q), program(nullptr), kernel(nullptr),
+          outData(nullptr), d_holder(gRandomSeed), testData(data_size),
+          kernel_source(src), max_workgroup_size(0)
+    {}
+
+    virtual cl_int SetUp(const char *kernel_source)
```

Comment:
Something doesn't seem right here - we have a `kernel_source` member variables and a `kernel_source` function argument.  Do we need both?

Reply:
Done, param renamed

Reply:
These tests are very difficult to understand and maintain currently, so I think we need to rethink how these tests are structured.  The parameter name was just a symptom.  Here are a few concrete suggestions:

1. I strongly suggest either leaving the existing `work_item_functions` test alone, or biting the bullet and refactoring it completely.  Currently, it appears to be split between keeping the same high-level test logic and moving to a test "fixture", which is then preventing the new tests from being implemented cleanly.
2. Having `hardcoded` as a template parameter is confusing.  It is used in exactly one place, to setup the right kernel string for the test.  Could we just pass in the right kernel string instead, from the start?  If not, could we pass it to the constructor to setup the kernel string properly and then be done with it?
3. Having a `kernel_source` member variable and a `src` argument passed to the `Setup` function is confusing, and I still don't know why we have both.  To make matters worse, in some cases we just pass `kernel_source` to `Setup`, and in some cases we don't.
4. Why is there both a constructor and a `Setup` function?  Could we move some of the functionality out of `Setup` into the constructor?  If not, why not?
5. The `work_item_functions` test does not have a `Validate` function, but the `work_item_functions_out_of_range` tests do.  Why is this?